// @databases/pg-data-type-id/lib/index.js
"use strict";
// autogenerated by test suite of pg-schema

Object.defineProperty(exports, "__esModule", { value: true });
var PgDataTypeID;
(function (PgDataTypeID) {
  // === Boolean ===
  /**
   * boolean, 'true'/'false'
   */
  PgDataTypeID[PgDataTypeID["bool"] = 16] = "bool";
  // === UserDefined ===
  /**
   * variable-length string, binary values escaped
   */
  PgDataTypeID[PgDataTypeID["bytea"] = 17] = "bytea";
  /**
   * (block, offset), physical location of tuple
   */
  PgDataTypeID[PgDataTypeID["tid"] = 27] = "tid";
  /**
   * transaction id
   */
  PgDataTypeID[PgDataTypeID["xid"] = 28] = "xid";
  /**
   * command identifier type, sequence in transaction id
   */
  PgDataTypeID[PgDataTypeID["cid"] = 29] = "cid";
  PgDataTypeID[PgDataTypeID["json"] = 114] = "json";
  /**
   * XML content
   */
  PgDataTypeID[PgDataTypeID["xml"] = 142] = "xml";
  /**
   * storage manager
   */
  PgDataTypeID[PgDataTypeID["smgr"] = 210] = "smgr";
  /**
   * XX:XX:XX:XX:XX:XX, MAC address
   */
  PgDataTypeID[PgDataTypeID["macaddr"] = 829] = "macaddr";
  /**
   * XX:XX:XX:XX:XX:XX:XX:XX, MAC address
   */
  PgDataTypeID[PgDataTypeID["macaddr8"] = 774] = "macaddr8";
  /**
   * access control list
   */
  PgDataTypeID[PgDataTypeID["aclitem"] = 1033] = "aclitem";
  /**
   * reference to cursor (portal name)
   */
  PgDataTypeID[PgDataTypeID["refcursor"] = 1790] = "refcursor";
  /**
   * UUID datatype
   */
  PgDataTypeID[PgDataTypeID["uuid"] = 2950] = "uuid";
  /**
   * PostgreSQL LSN datatype
   */
  PgDataTypeID[PgDataTypeID["pg_lsn"] = 3220] = "pg_lsn";
  /**
   * text representation for text search
   */
  PgDataTypeID[PgDataTypeID["tsvector"] = 3614] = "tsvector";
  /**
   * GiST index internal text representation for text search
   */
  PgDataTypeID[PgDataTypeID["gtsvector"] = 3642] = "gtsvector";
  /**
   * query representation for text search
   */
  PgDataTypeID[PgDataTypeID["tsquery"] = 3615] = "tsquery";
  /**
   * Binary JSON
   */
  PgDataTypeID[PgDataTypeID["jsonb"] = 3802] = "jsonb";
  /**
   * txid snapshot
   */
  PgDataTypeID[PgDataTypeID["txid_snapshot"] = 2970] = "txid_snapshot";
  // === String ===
  /**
   * single character
   */
  PgDataTypeID[PgDataTypeID["char"] = 18] = "char";
  /**
   * 63-byte type for storing system identifiers
   */
  PgDataTypeID[PgDataTypeID["name"] = 19] = "name";
  /**
   * variable-length string, no limit specified
   */
  PgDataTypeID[PgDataTypeID["text"] = 25] = "text";
  /**
   * string representing an internal node tree
   */
  PgDataTypeID[PgDataTypeID["pg_node_tree"] = 194] = "pg_node_tree";
  /**
   * multivariate ndistinct coefficients
   */
  PgDataTypeID[PgDataTypeID["pg_ndistinct"] = 3361] = "pg_ndistinct";
  /**
   * multivariate dependencies
   */
  PgDataTypeID[PgDataTypeID["pg_dependencies"] = 3402] = "pg_dependencies";
  /**
   * char(length), blank-padded string, fixed storage length
   */
  PgDataTypeID[PgDataTypeID["bpchar"] = 1042] = "bpchar";
  /**
   * varchar(length), non-blank-padded string, variable storage length
   */
  PgDataTypeID[PgDataTypeID["varchar"] = 1043] = "varchar";
  // === Numeric ===
  /**
   * ~18 digit integer, 8-byte storage
   */
  PgDataTypeID[PgDataTypeID["int8"] = 20] = "int8";
  /**
   * -32 thousand to 32 thousand, 2-byte storage
   */
  PgDataTypeID[PgDataTypeID["int2"] = 21] = "int2";
  /**
   * -2 billion to 2 billion integer, 4-byte storage
   */
  PgDataTypeID[PgDataTypeID["int4"] = 23] = "int4";
  /**
   * registered procedure
   */
  PgDataTypeID[PgDataTypeID["regproc"] = 24] = "regproc";
  /**
   * object identifier(oid), maximum 4 billion
   */
  PgDataTypeID[PgDataTypeID["oid"] = 26] = "oid";
  /**
   * single-precision floating point number, 4-byte storage
   */
  PgDataTypeID[PgDataTypeID["float4"] = 700] = "float4";
  /**
   * double-precision floating point number, 8-byte storage
   */
  PgDataTypeID[PgDataTypeID["float8"] = 701] = "float8";
  /**
   * monetary amounts, $d,ddd.cc
   */
  PgDataTypeID[PgDataTypeID["money"] = 790] = "money";
  /**
   * numeric(precision, decimal), arbitrary precision number
   */
  PgDataTypeID[PgDataTypeID["numeric"] = 1700] = "numeric";
  /**
   * registered procedure (with args)
   */
  PgDataTypeID[PgDataTypeID["regprocedure"] = 2202] = "regprocedure";
  /**
   * registered operator
   */
  PgDataTypeID[PgDataTypeID["regoper"] = 2203] = "regoper";
  /**
   * registered operator (with args)
   */
  PgDataTypeID[PgDataTypeID["regoperator"] = 2204] = "regoperator";
  /**
   * registered class
   */
  PgDataTypeID[PgDataTypeID["regclass"] = 2205] = "regclass";
  /**
   * registered type
   */
  PgDataTypeID[PgDataTypeID["regtype"] = 2206] = "regtype";
  /**
   * registered role
   */
  PgDataTypeID[PgDataTypeID["regrole"] = 4096] = "regrole";
  /**
   * registered namespace
   */
  PgDataTypeID[PgDataTypeID["regnamespace"] = 4089] = "regnamespace";
  /**
   * registered text search configuration
   */
  PgDataTypeID[PgDataTypeID["regconfig"] = 3734] = "regconfig";
  /**
   * registered text search dictionary
   */
  PgDataTypeID[PgDataTypeID["regdictionary"] = 3769] = "regdictionary";
  // === Array ===
  /**
   * array of int2, used in system tables
   *
   * Array<int2>
   */
  PgDataTypeID[PgDataTypeID["int2vector"] = 22] = "int2vector";
  /**
   * array of oids, used in system tables
   *
   * Array<oid>
   */
  PgDataTypeID[PgDataTypeID["oidvector"] = 30] = "oidvector";
  /**
   * Array<xml>
   */
  PgDataTypeID[PgDataTypeID["_xml"] = 143] = "_xml";
  /**
   * Array<json>
   */
  PgDataTypeID[PgDataTypeID["_json"] = 199] = "_json";
  /**
   * Array<line>
   */
  PgDataTypeID[PgDataTypeID["_line"] = 629] = "_line";
  /**
   * Array<circle>
   */
  PgDataTypeID[PgDataTypeID["_circle"] = 719] = "_circle";
  /**
   * Array<money>
   */
  PgDataTypeID[PgDataTypeID["_money"] = 791] = "_money";
  /**
   * Array<bool>
   */
  PgDataTypeID[PgDataTypeID["_bool"] = 1000] = "_bool";
  /**
   * Array<bytea>
   */
  PgDataTypeID[PgDataTypeID["_bytea"] = 1001] = "_bytea";
  /**
   * Array<char>
   */
  PgDataTypeID[PgDataTypeID["_char"] = 1002] = "_char";
  /**
   * Array<name>
   */
  PgDataTypeID[PgDataTypeID["_name"] = 1003] = "_name";
  /**
   * Array<int2>
   */
  PgDataTypeID[PgDataTypeID["_int2"] = 1005] = "_int2";
  /**
   * Array<int2vector>
   */
  PgDataTypeID[PgDataTypeID["_int2vector"] = 1006] = "_int2vector";
  /**
   * Array<int4>
   */
  PgDataTypeID[PgDataTypeID["_int4"] = 1007] = "_int4";
  /**
   * Array<regproc>
   */
  PgDataTypeID[PgDataTypeID["_regproc"] = 1008] = "_regproc";
  /**
   * Array<text>
   */
  PgDataTypeID[PgDataTypeID["_text"] = 1009] = "_text";
  /**
   * Array<oid>
   */
  PgDataTypeID[PgDataTypeID["_oid"] = 1028] = "_oid";
  /**
   * Array<tid>
   */
  PgDataTypeID[PgDataTypeID["_tid"] = 1010] = "_tid";
  /**
   * Array<xid>
   */
  PgDataTypeID[PgDataTypeID["_xid"] = 1011] = "_xid";
  /**
   * Array<cid>
   */
  PgDataTypeID[PgDataTypeID["_cid"] = 1012] = "_cid";
  /**
   * Array<oidvector>
   */
  PgDataTypeID[PgDataTypeID["_oidvector"] = 1013] = "_oidvector";
  /**
   * Array<bpchar>
   */
  PgDataTypeID[PgDataTypeID["_bpchar"] = 1014] = "_bpchar";
  /**
   * Array<varchar>
   */
  PgDataTypeID[PgDataTypeID["_varchar"] = 1015] = "_varchar";
  /**
   * Array<int8>
   */
  PgDataTypeID[PgDataTypeID["_int8"] = 1016] = "_int8";
  /**
   * Array<point>
   */
  PgDataTypeID[PgDataTypeID["_point"] = 1017] = "_point";
  /**
   * Array<lseg>
   */
  PgDataTypeID[PgDataTypeID["_lseg"] = 1018] = "_lseg";
  /**
   * Array<path>
   */
  PgDataTypeID[PgDataTypeID["_path"] = 1019] = "_path";
  /**
   * Array<box>
   */
  PgDataTypeID[PgDataTypeID["_box"] = 1020] = "_box";
  /**
   * Array<float4>
   */
  PgDataTypeID[PgDataTypeID["_float4"] = 1021] = "_float4";
  /**
   * Array<float8>
   */
  PgDataTypeID[PgDataTypeID["_float8"] = 1022] = "_float8";
  /**
   * Array<abstime>
   */
  PgDataTypeID[PgDataTypeID["_abstime"] = 1023] = "_abstime";
  /**
   * Array<reltime>
   */
  PgDataTypeID[PgDataTypeID["_reltime"] = 1024] = "_reltime";
  /**
   * Array<tinterval>
   */
  PgDataTypeID[PgDataTypeID["_tinterval"] = 1025] = "_tinterval";
  /**
   * Array<polygon>
   */
  PgDataTypeID[PgDataTypeID["_polygon"] = 1027] = "_polygon";
  /**
   * Array<aclitem>
   */
  PgDataTypeID[PgDataTypeID["_aclitem"] = 1034] = "_aclitem";
  /**
   * Array<macaddr>
   */
  PgDataTypeID[PgDataTypeID["_macaddr"] = 1040] = "_macaddr";
  /**
   * Array<macaddr8>
   */
  PgDataTypeID[PgDataTypeID["_macaddr8"] = 775] = "_macaddr8";
  /**
   * Array<inet>
   */
  PgDataTypeID[PgDataTypeID["_inet"] = 1041] = "_inet";
  /**
   * Array<cidr>
   */
  PgDataTypeID[PgDataTypeID["_cidr"] = 651] = "_cidr";
  /**
   * Array<cstring>
   */
  PgDataTypeID[PgDataTypeID["_cstring"] = 1263] = "_cstring";
  /**
   * Array<timestamp>
   */
  PgDataTypeID[PgDataTypeID["_timestamp"] = 1115] = "_timestamp";
  /**
   * Array<date>
   */
  PgDataTypeID[PgDataTypeID["_date"] = 1182] = "_date";
  /**
   * Array<time>
   */
  PgDataTypeID[PgDataTypeID["_time"] = 1183] = "_time";
  /**
   * Array<timestamptz>
   */
  PgDataTypeID[PgDataTypeID["_timestamptz"] = 1185] = "_timestamptz";
  /**
   * Array<interval>
   */
  PgDataTypeID[PgDataTypeID["_interval"] = 1187] = "_interval";
  /**
   * Array<numeric>
   */
  PgDataTypeID[PgDataTypeID["_numeric"] = 1231] = "_numeric";
  /**
   * Array<timetz>
   */
  PgDataTypeID[PgDataTypeID["_timetz"] = 1270] = "_timetz";
  /**
   * Array<bit>
   */
  PgDataTypeID[PgDataTypeID["_bit"] = 1561] = "_bit";
  /**
   * Array<varbit>
   */
  PgDataTypeID[PgDataTypeID["_varbit"] = 1563] = "_varbit";
  /**
   * Array<refcursor>
   */
  PgDataTypeID[PgDataTypeID["_refcursor"] = 2201] = "_refcursor";
  /**
   * Array<regprocedure>
   */
  PgDataTypeID[PgDataTypeID["_regprocedure"] = 2207] = "_regprocedure";
  /**
   * Array<regoper>
   */
  PgDataTypeID[PgDataTypeID["_regoper"] = 2208] = "_regoper";
  /**
   * Array<regoperator>
   */
  PgDataTypeID[PgDataTypeID["_regoperator"] = 2209] = "_regoperator";
  /**
   * Array<regclass>
   */
  PgDataTypeID[PgDataTypeID["_regclass"] = 2210] = "_regclass";
  /**
   * Array<regtype>
   */
  PgDataTypeID[PgDataTypeID["_regtype"] = 2211] = "_regtype";
  /**
   * Array<regrole>
   */
  PgDataTypeID[PgDataTypeID["_regrole"] = 4097] = "_regrole";
  /**
   * Array<regnamespace>
   */
  PgDataTypeID[PgDataTypeID["_regnamespace"] = 4090] = "_regnamespace";
  /**
   * Array<uuid>
   */
  PgDataTypeID[PgDataTypeID["_uuid"] = 2951] = "_uuid";
  /**
   * Array<pg_lsn>
   */
  PgDataTypeID[PgDataTypeID["_pg_lsn"] = 3221] = "_pg_lsn";
  /**
   * Array<tsvector>
   */
  PgDataTypeID[PgDataTypeID["_tsvector"] = 3643] = "_tsvector";
  /**
   * Array<gtsvector>
   */
  PgDataTypeID[PgDataTypeID["_gtsvector"] = 3644] = "_gtsvector";
  /**
   * Array<tsquery>
   */
  PgDataTypeID[PgDataTypeID["_tsquery"] = 3645] = "_tsquery";
  /**
   * Array<regconfig>
   */
  PgDataTypeID[PgDataTypeID["_regconfig"] = 3735] = "_regconfig";
  /**
   * Array<regdictionary>
   */
  PgDataTypeID[PgDataTypeID["_regdictionary"] = 3770] = "_regdictionary";
  /**
   * Array<jsonb>
   */
  PgDataTypeID[PgDataTypeID["_jsonb"] = 3807] = "_jsonb";
  /**
   * Array<txid_snapshot>
   */
  PgDataTypeID[PgDataTypeID["_txid_snapshot"] = 2949] = "_txid_snapshot";
  /**
   * Array<int4range>
   */
  PgDataTypeID[PgDataTypeID["_int4range"] = 3905] = "_int4range";
  /**
   * Array<numrange>
   */
  PgDataTypeID[PgDataTypeID["_numrange"] = 3907] = "_numrange";
  /**
   * Array<tsrange>
   */
  PgDataTypeID[PgDataTypeID["_tsrange"] = 3909] = "_tsrange";
  /**
   * Array<tstzrange>
   */
  PgDataTypeID[PgDataTypeID["_tstzrange"] = 3911] = "_tstzrange";
  /**
   * Array<daterange>
   */
  PgDataTypeID[PgDataTypeID["_daterange"] = 3913] = "_daterange";
  /**
   * Array<int8range>
   */
  PgDataTypeID[PgDataTypeID["_int8range"] = 3927] = "_int8range";
  // === Composite ===
  PgDataTypeID[PgDataTypeID["pg_type"] = 71] = "pg_type";
  PgDataTypeID[PgDataTypeID["pg_attribute"] = 75] = "pg_attribute";
  PgDataTypeID[PgDataTypeID["pg_proc"] = 81] = "pg_proc";
  PgDataTypeID[PgDataTypeID["pg_class"] = 83] = "pg_class";
  PgDataTypeID[PgDataTypeID["pg_attrdef"] = 10000] = "pg_attrdef";
  PgDataTypeID[PgDataTypeID["pg_constraint"] = 10001] = "pg_constraint";
  PgDataTypeID[PgDataTypeID["pg_inherits"] = 10002] = "pg_inherits";
  PgDataTypeID[PgDataTypeID["pg_index"] = 10003] = "pg_index";
  PgDataTypeID[PgDataTypeID["pg_operator"] = 10004] = "pg_operator";
  PgDataTypeID[PgDataTypeID["pg_opfamily"] = 10005] = "pg_opfamily";
  PgDataTypeID[PgDataTypeID["pg_opclass"] = 10006] = "pg_opclass";
  PgDataTypeID[PgDataTypeID["pg_am"] = 10130] = "pg_am";
  PgDataTypeID[PgDataTypeID["pg_amop"] = 10131] = "pg_amop";
  PgDataTypeID[PgDataTypeID["pg_amproc"] = 10841] = "pg_amproc";
  PgDataTypeID[PgDataTypeID["pg_language"] = 11253] = "pg_language";
  PgDataTypeID[PgDataTypeID["pg_largeobject_metadata"] = 11254] = "pg_largeobject_metadata";
  PgDataTypeID[PgDataTypeID["pg_largeobject"] = 11255] = "pg_largeobject";
  PgDataTypeID[PgDataTypeID["pg_aggregate"] = 11256] = "pg_aggregate";
  PgDataTypeID[PgDataTypeID["pg_statistic_ext"] = 11257] = "pg_statistic_ext";
  PgDataTypeID[PgDataTypeID["pg_statistic"] = 11258] = "pg_statistic";
  PgDataTypeID[PgDataTypeID["pg_rewrite"] = 11259] = "pg_rewrite";
  PgDataTypeID[PgDataTypeID["pg_trigger"] = 11260] = "pg_trigger";
  PgDataTypeID[PgDataTypeID["pg_event_trigger"] = 11261] = "pg_event_trigger";
  PgDataTypeID[PgDataTypeID["pg_description"] = 11262] = "pg_description";
  PgDataTypeID[PgDataTypeID["pg_cast"] = 11263] = "pg_cast";
  PgDataTypeID[PgDataTypeID["pg_enum"] = 11483] = "pg_enum";
  PgDataTypeID[PgDataTypeID["pg_namespace"] = 11484] = "pg_namespace";
  PgDataTypeID[PgDataTypeID["pg_conversion"] = 11485] = "pg_conversion";
  PgDataTypeID[PgDataTypeID["pg_depend"] = 11486] = "pg_depend";
  PgDataTypeID[PgDataTypeID["pg_database"] = 1248] = "pg_database";
  PgDataTypeID[PgDataTypeID["pg_db_role_setting"] = 11487] = "pg_db_role_setting";
  PgDataTypeID[PgDataTypeID["pg_tablespace"] = 11488] = "pg_tablespace";
  PgDataTypeID[PgDataTypeID["pg_pltemplate"] = 11489] = "pg_pltemplate";
  PgDataTypeID[PgDataTypeID["pg_authid"] = 2842] = "pg_authid";
  PgDataTypeID[PgDataTypeID["pg_auth_members"] = 2843] = "pg_auth_members";
  PgDataTypeID[PgDataTypeID["pg_shdepend"] = 11490] = "pg_shdepend";
  PgDataTypeID[PgDataTypeID["pg_shdescription"] = 11491] = "pg_shdescription";
  PgDataTypeID[PgDataTypeID["pg_ts_config"] = 11492] = "pg_ts_config";
  PgDataTypeID[PgDataTypeID["pg_ts_config_map"] = 11493] = "pg_ts_config_map";
  PgDataTypeID[PgDataTypeID["pg_ts_dict"] = 11494] = "pg_ts_dict";
  PgDataTypeID[PgDataTypeID["pg_ts_parser"] = 11495] = "pg_ts_parser";
  PgDataTypeID[PgDataTypeID["pg_ts_template"] = 11496] = "pg_ts_template";
  PgDataTypeID[PgDataTypeID["pg_extension"] = 11497] = "pg_extension";
  PgDataTypeID[PgDataTypeID["pg_foreign_data_wrapper"] = 11498] = "pg_foreign_data_wrapper";
  PgDataTypeID[PgDataTypeID["pg_foreign_server"] = 11499] = "pg_foreign_server";
  PgDataTypeID[PgDataTypeID["pg_user_mapping"] = 11500] = "pg_user_mapping";
  PgDataTypeID[PgDataTypeID["pg_foreign_table"] = 11501] = "pg_foreign_table";
  PgDataTypeID[PgDataTypeID["pg_policy"] = 11502] = "pg_policy";
  PgDataTypeID[PgDataTypeID["pg_replication_origin"] = 11503] = "pg_replication_origin";
  PgDataTypeID[PgDataTypeID["pg_default_acl"] = 11504] = "pg_default_acl";
  PgDataTypeID[PgDataTypeID["pg_init_privs"] = 11505] = "pg_init_privs";
  PgDataTypeID[PgDataTypeID["pg_seclabel"] = 11506] = "pg_seclabel";
  PgDataTypeID[PgDataTypeID["pg_shseclabel"] = 4066] = "pg_shseclabel";
  PgDataTypeID[PgDataTypeID["pg_collation"] = 11507] = "pg_collation";
  PgDataTypeID[PgDataTypeID["pg_partitioned_table"] = 11508] = "pg_partitioned_table";
  PgDataTypeID[PgDataTypeID["pg_range"] = 11509] = "pg_range";
  PgDataTypeID[PgDataTypeID["pg_transform"] = 11510] = "pg_transform";
  PgDataTypeID[PgDataTypeID["pg_sequence"] = 11511] = "pg_sequence";
  PgDataTypeID[PgDataTypeID["pg_publication"] = 11512] = "pg_publication";
  PgDataTypeID[PgDataTypeID["pg_publication_rel"] = 11513] = "pg_publication_rel";
  PgDataTypeID[PgDataTypeID["pg_subscription"] = 6101] = "pg_subscription";
  PgDataTypeID[PgDataTypeID["pg_subscription_rel"] = 11514] = "pg_subscription_rel";
  PgDataTypeID[PgDataTypeID["pg_roles"] = 11528] = "pg_roles";
  PgDataTypeID[PgDataTypeID["pg_shadow"] = 11532] = "pg_shadow";
  PgDataTypeID[PgDataTypeID["pg_group"] = 11536] = "pg_group";
  PgDataTypeID[PgDataTypeID["pg_user"] = 11539] = "pg_user";
  PgDataTypeID[PgDataTypeID["pg_policies"] = 11542] = "pg_policies";
  PgDataTypeID[PgDataTypeID["pg_rules"] = 11546] = "pg_rules";
  PgDataTypeID[PgDataTypeID["pg_views"] = 11550] = "pg_views";
  PgDataTypeID[PgDataTypeID["pg_tables"] = 11554] = "pg_tables";
  PgDataTypeID[PgDataTypeID["pg_matviews"] = 11558] = "pg_matviews";
  PgDataTypeID[PgDataTypeID["pg_indexes"] = 11562] = "pg_indexes";
  PgDataTypeID[PgDataTypeID["pg_sequences"] = 11566] = "pg_sequences";
  PgDataTypeID[PgDataTypeID["pg_stats"] = 11570] = "pg_stats";
  PgDataTypeID[PgDataTypeID["pg_publication_tables"] = 11574] = "pg_publication_tables";
  PgDataTypeID[PgDataTypeID["pg_locks"] = 11578] = "pg_locks";
  PgDataTypeID[PgDataTypeID["pg_cursors"] = 11581] = "pg_cursors";
  PgDataTypeID[PgDataTypeID["pg_available_extensions"] = 11584] = "pg_available_extensions";
  PgDataTypeID[PgDataTypeID["pg_available_extension_versions"] = 11587] = "pg_available_extension_versions";
  PgDataTypeID[PgDataTypeID["pg_prepared_xacts"] = 11590] = "pg_prepared_xacts";
  PgDataTypeID[PgDataTypeID["pg_prepared_statements"] = 11594] = "pg_prepared_statements";
  PgDataTypeID[PgDataTypeID["pg_seclabels"] = 11597] = "pg_seclabels";
  PgDataTypeID[PgDataTypeID["pg_settings"] = 11601] = "pg_settings";
  PgDataTypeID[PgDataTypeID["pg_file_settings"] = 11606] = "pg_file_settings";
  PgDataTypeID[PgDataTypeID["pg_hba_file_rules"] = 11609] = "pg_hba_file_rules";
  PgDataTypeID[PgDataTypeID["pg_timezone_abbrevs"] = 11612] = "pg_timezone_abbrevs";
  PgDataTypeID[PgDataTypeID["pg_timezone_names"] = 11615] = "pg_timezone_names";
  PgDataTypeID[PgDataTypeID["pg_config"] = 11618] = "pg_config";
  PgDataTypeID[PgDataTypeID["pg_stat_all_tables"] = 11621] = "pg_stat_all_tables";
  PgDataTypeID[PgDataTypeID["pg_stat_xact_all_tables"] = 11625] = "pg_stat_xact_all_tables";
  PgDataTypeID[PgDataTypeID["pg_stat_sys_tables"] = 11629] = "pg_stat_sys_tables";
  PgDataTypeID[PgDataTypeID["pg_stat_xact_sys_tables"] = 11633] = "pg_stat_xact_sys_tables";
  PgDataTypeID[PgDataTypeID["pg_stat_user_tables"] = 11636] = "pg_stat_user_tables";
  PgDataTypeID[PgDataTypeID["pg_stat_xact_user_tables"] = 11640] = "pg_stat_xact_user_tables";
  PgDataTypeID[PgDataTypeID["pg_statio_all_tables"] = 11643] = "pg_statio_all_tables";
  PgDataTypeID[PgDataTypeID["pg_statio_sys_tables"] = 11647] = "pg_statio_sys_tables";
  PgDataTypeID[PgDataTypeID["pg_statio_user_tables"] = 11650] = "pg_statio_user_tables";
  PgDataTypeID[PgDataTypeID["pg_stat_all_indexes"] = 11653] = "pg_stat_all_indexes";
  PgDataTypeID[PgDataTypeID["pg_stat_sys_indexes"] = 11657] = "pg_stat_sys_indexes";
  PgDataTypeID[PgDataTypeID["pg_stat_user_indexes"] = 11660] = "pg_stat_user_indexes";
  PgDataTypeID[PgDataTypeID["pg_statio_all_indexes"] = 11663] = "pg_statio_all_indexes";
  PgDataTypeID[PgDataTypeID["pg_statio_sys_indexes"] = 11667] = "pg_statio_sys_indexes";
  PgDataTypeID[PgDataTypeID["pg_statio_user_indexes"] = 11670] = "pg_statio_user_indexes";
  PgDataTypeID[PgDataTypeID["pg_statio_all_sequences"] = 11673] = "pg_statio_all_sequences";
  PgDataTypeID[PgDataTypeID["pg_statio_sys_sequences"] = 11677] = "pg_statio_sys_sequences";
  PgDataTypeID[PgDataTypeID["pg_statio_user_sequences"] = 11680] = "pg_statio_user_sequences";
  PgDataTypeID[PgDataTypeID["pg_stat_activity"] = 11683] = "pg_stat_activity";
  PgDataTypeID[PgDataTypeID["pg_stat_replication"] = 11687] = "pg_stat_replication";
  PgDataTypeID[PgDataTypeID["pg_stat_wal_receiver"] = 11691] = "pg_stat_wal_receiver";
  PgDataTypeID[PgDataTypeID["pg_stat_subscription"] = 11694] = "pg_stat_subscription";
  PgDataTypeID[PgDataTypeID["pg_stat_ssl"] = 11697] = "pg_stat_ssl";
  PgDataTypeID[PgDataTypeID["pg_replication_slots"] = 11700] = "pg_replication_slots";
  PgDataTypeID[PgDataTypeID["pg_stat_database"] = 11704] = "pg_stat_database";
  PgDataTypeID[PgDataTypeID["pg_stat_database_conflicts"] = 11707] = "pg_stat_database_conflicts";
  PgDataTypeID[PgDataTypeID["pg_stat_user_functions"] = 11710] = "pg_stat_user_functions";
  PgDataTypeID[PgDataTypeID["pg_stat_xact_user_functions"] = 11714] = "pg_stat_xact_user_functions";
  PgDataTypeID[PgDataTypeID["pg_stat_archiver"] = 11718] = "pg_stat_archiver";
  PgDataTypeID[PgDataTypeID["pg_stat_bgwriter"] = 11721] = "pg_stat_bgwriter";
  PgDataTypeID[PgDataTypeID["pg_stat_progress_vacuum"] = 11724] = "pg_stat_progress_vacuum";
  PgDataTypeID[PgDataTypeID["pg_user_mappings"] = 11728] = "pg_user_mappings";
  PgDataTypeID[PgDataTypeID["pg_replication_origin_status"] = 11732] = "pg_replication_origin_status";
  // === PseudoTypes ===
  /**
   * internal type for passing CollectedCommand
   */
  PgDataTypeID[PgDataTypeID["pg_ddl_command"] = 32] = "pg_ddl_command";
  PgDataTypeID[PgDataTypeID["record"] = 2249] = "record";
  PgDataTypeID[PgDataTypeID["_record"] = 2287] = "_record";
  PgDataTypeID[PgDataTypeID["cstring"] = 2275] = "cstring";
  PgDataTypeID[PgDataTypeID["any"] = 2276] = "any";
  PgDataTypeID[PgDataTypeID["anyarray"] = 2277] = "anyarray";
  PgDataTypeID[PgDataTypeID["void"] = 2278] = "void";
  PgDataTypeID[PgDataTypeID["trigger"] = 2279] = "trigger";
  PgDataTypeID[PgDataTypeID["event_trigger"] = 3838] = "event_trigger";
  PgDataTypeID[PgDataTypeID["language_handler"] = 2280] = "language_handler";
  PgDataTypeID[PgDataTypeID["internal"] = 2281] = "internal";
  PgDataTypeID[PgDataTypeID["opaque"] = 2282] = "opaque";
  PgDataTypeID[PgDataTypeID["anyelement"] = 2283] = "anyelement";
  PgDataTypeID[PgDataTypeID["anynonarray"] = 2776] = "anynonarray";
  PgDataTypeID[PgDataTypeID["anyenum"] = 3500] = "anyenum";
  PgDataTypeID[PgDataTypeID["fdw_handler"] = 3115] = "fdw_handler";
  PgDataTypeID[PgDataTypeID["index_am_handler"] = 325] = "index_am_handler";
  PgDataTypeID[PgDataTypeID["tsm_handler"] = 3310] = "tsm_handler";
  PgDataTypeID[PgDataTypeID["anyrange"] = 3831] = "anyrange";
  // === Geometric ===
  /**
   * geometric point '(x, y)'
   */
  PgDataTypeID[PgDataTypeID["point"] = 600] = "point";
  /**
   * geometric line segment '(pt1,pt2)'
   */
  PgDataTypeID[PgDataTypeID["lseg"] = 601] = "lseg";
  /**
   * geometric path '(pt1,...)'
   */
  PgDataTypeID[PgDataTypeID["path"] = 602] = "path";
  /**
   * geometric box '(lower left,upper right)'
   */
  PgDataTypeID[PgDataTypeID["box"] = 603] = "box";
  /**
   * geometric polygon '(pt1,...)'
   */
  PgDataTypeID[PgDataTypeID["polygon"] = 604] = "polygon";
  /**
   * geometric line
   */
  PgDataTypeID[PgDataTypeID["line"] = 628] = "line";
  /**
   * geometric circle '(center,radius)'
   */
  PgDataTypeID[PgDataTypeID["circle"] = 718] = "circle";
  // === DateTime ===
  /**
   * absolute, limited-range date and time (Unix system time)
   */
  PgDataTypeID[PgDataTypeID["abstime"] = 702] = "abstime";
  /**
   * date
   */
  PgDataTypeID[PgDataTypeID["date"] = 1082] = "date";
  /**
   * time of day
   */
  PgDataTypeID[PgDataTypeID["time"] = 1083] = "time";
  /**
   * date and time
   */
  PgDataTypeID[PgDataTypeID["timestamp"] = 1114] = "timestamp";
  /**
   * date and time with time zone
   */
  PgDataTypeID[PgDataTypeID["timestamptz"] = 1184] = "timestamptz";
  /**
   * time of day with time zone
   */
  PgDataTypeID[PgDataTypeID["timetz"] = 1266] = "timetz";
  // === Timespan ===
  /**
   * relative, limited-range time interval (Unix delta time)
   */
  PgDataTypeID[PgDataTypeID["reltime"] = 703] = "reltime";
  /**
   * (abstime,abstime), time interval
   */
  PgDataTypeID[PgDataTypeID["tinterval"] = 704] = "tinterval";
  /**
   * @ <number> <units>, time interval
   */
  PgDataTypeID[PgDataTypeID["interval"] = 1186] = "interval";
  // === Unknown ===
  PgDataTypeID[PgDataTypeID["unknown"] = 705] = "unknown";
  // === NetworkAddress ===
  /**
   * IP address/netmask, host address, netmask optional
   */
  PgDataTypeID[PgDataTypeID["inet"] = 869] = "inet";
  /**
   * network IP address/netmask, network address
   */
  PgDataTypeID[PgDataTypeID["cidr"] = 650] = "cidr";
  // === BitString ===
  /**
   * fixed-length bit string
   */
  PgDataTypeID[PgDataTypeID["bit"] = 1560] = "bit";
  /**
   * variable-length bit string
   */
  PgDataTypeID[PgDataTypeID["varbit"] = 1562] = "varbit";
  // === Range ===
  /**
   * range of integers
   */
  PgDataTypeID[PgDataTypeID["int4range"] = 3904] = "int4range";
  /**
   * range of numerics
   */
  PgDataTypeID[PgDataTypeID["numrange"] = 3906] = "numrange";
  /**
   * range of timestamps without time zone
   */
  PgDataTypeID[PgDataTypeID["tsrange"] = 3908] = "tsrange";
  /**
   * range of timestamps with time zone
   */
  PgDataTypeID[PgDataTypeID["tstzrange"] = 3910] = "tstzrange";
  /**
   * range of dates
   */
  PgDataTypeID[PgDataTypeID["daterange"] = 3912] = "daterange";
  /**
   * range of bigints
   */
  PgDataTypeID[PgDataTypeID["int8range"] = 3926] = "int8range";
})(PgDataTypeID || (PgDataTypeID = {}));
exports.default = PgDataTypeID;
module.exports = PgDataTypeID;
module.exports.default = PgDataTypeID;
